### ISSUE TITLE
Clarification Appendix A.3.

### ISSUE THREAD

**Issue by StefanHri** [lake-wg/edhoc#257]
Hi,

I don't understand this very good: 
> "For the default case when the
   server is Responder, the prepended connection identifier is C_R, and
   C_I if the server is Initiator.  If message_1 is sent to the server,
   the CBOR simple value "true" (0xf5) is sent in its place (given that
   the server has not selected C_R yet)."

In the case where:
* the initiator is a client and
* the responder is a server
Do we need to replace C_I with 0xf5 in message_1??? 

@AleksanderDemianowski

**Comment by gselander** [comment]
No, this is only referring to the *prepended* connection identifiers. 

The prepended connection identifiers are only used when the session can't be inferred from the previous message. The first message is a special case since there is no previous message, which is why "true" is used. 

Maybe the "in its place" is not sufficiently clear?

**Comment by AleksanderDemianowski** [comment]
@StefanHri just for clarifiaction, @gselander please correct me if needed. Assuming the responder is a server, the first send data it's a message from client to server. This message is a concatenation of 1B (true - 0xf5) and EDHOC message_1. EDHOC message_1 includes C_I inside and this is not changing (is not repclaced). 
Always client (initiator) sends data like: (0xf5) || (METHOD, SUITES_I, G_X, C_I, EAD_1 ), where || is a symbol of concatenation.
This is what we can see as the first message on [Figure 10 in A.3. chapter.](https://tools.ietf.org/id/draft-ietf-lake-edhoc-12.html#fig-coap1)

**Comment by gselander** [comment]
@AleksanderDemianowski Yes that is correct. 

Here is some new text replacing the existing description in an attempt to clarify further:

OLD
> In order to correlate a message received from a client to a message previously sent by the server, messages sent by the client are prepended with the CBOR serialization of the connection identifier which the server has chosen.  For the default case when the server is Responder, the prepended connection identifier is C_R, and C_I if the server is Initiator. If message_1 is sent to the server, the CBOR simple value "true" (0xf5) is sent in its place (given that the server has not selected C_R yet).


NEW
> In order to correlate a message received from a client to a message previously sent by the server in the same EDHOC session over CoAP, messages sent by the client are prepended with the CBOR serialization of the connection identifier which the server has chosen. This applies independently of if the CoAP server is Responder or Initiator. 

> * For the default case when the server is Responder, message_3 sent from the client is prepended with the identifier C_R. In this case message_1 is also sent by the client, and to indicate that this is a new EDHOC session it is prepended with a dummy identifier, the CBOR simple value "true" (0xf5), since the server has not selected C_R yet. See Figure 10.

> * In the case when the server is Initiator, message_2 sent from the client is prepended with the identifier C_I. See Figure 11.

> Note the prepended identifiers are only used to keep track of messages in an EDHOC session, and this does not impact the value of the connection identifiers within the EDHOC messages.

@StefanHri Is this more clear?

**Comment by gselander** [comment]
Some adaptations of the proposal above in PR #259.

**Comment by StefanHri** [comment]
Thanks @gselander  and @AleksanderDemianowski  for the explanations! For me the new text is now clear. 

However I have now a more generic question. For me correlation and protocol session maintenance are closely related. Ideally an EDHOC implementation should support several sessions running in parallel, i.e., several protocol states should be maintained. Ideally this state information should be not visible to the user. 

Appendix A.3 proposes two different mechanisms for corelation: 1) on the server side through the prepended connection identifier chosen by the server and 2) on the client side through the CoAP token.

For me this means that the mapping of messages to sessions depend on the transport protocol, which makes implementing a transport independent EDHOC library difficult. 

Is it not better to provide a mechanism where message can be correlated independently from the transport protocol? Obviously, this will increase the bytes on the wire, however then EDHOC will be truly independent form the transport which will allow generic EDHOC libraries to be used with different transports.

**Comment by gselander** [comment]
I think Stefan has a good point about providing an implementation independent of transport. Do others agree?

Even if the correlation problem is out of EDHOC, and a solution is specified for CoAP, there is a need for adding adaptation for any other transport. Can this be made more application friendly? Would a "wrapper" prepending connection ids on all message be a (non-optimized) solution which can be used for transport independence?

@chrysn @marco-tiloca-sics @stoprocent and others: What do you think 
(a) is a user-friendly solution, and
(b) should be stated in the draft.

**Comment by marco-tiloca-sics** [comment]
If any possible transport has to be covered (CoAP included), then prepending connection identifiers to all EDHOC messages is the only way I see to be entirely agnostic of the specifically used transport (even when not necessary since the transport properties allow to achieve correlation anyway).

However, on top of being not justified due to the transport properties ensuring correlation anyway, the resulting communication overhead can be unacceptable.

I am thinking especially of EDHOC message_2 transported with CoAP in the typical Client=Initiator workflow. The smallest possible actual message_2 would still be 45 bytes, but it will yield a message payload of at least 46 bytes on the wire due to the prepended (and useless) C_I, and we know 45 bytes to be the critical payload threshold which is possible to not exceed by leveraging the inclusion of connection identifiers only when necessary.

Since those prepended connection identifiers are not part of the actual EDHOC messages, the code building an EDHOC message based on the current protocol status can already be transport-agnostic.

For implementations of EDHOC that are expected to be used with multiple possible transports (e.g., as stand-alone libraries), I see the mentioned wrapper as an implementation approach where the payload to send on the wire is finalized starting from the just prepared EDHOC message. However, due to the reasons above, I think that at least the wrapper has to be transport-aware, and prepend a connection identifier or not in accordance with the correlation properties of the specifically used transport.

This may be worth stating in the draft, but I think mostly as notes for implementors.

**Comment by chrysn** [comment]
Such a wrapper might need to provide all kinds of things -- essentially the 3.4 "shopping list" (of which message correlation is but one point). (And one might also add "framing" as in "determining the message's length" to the list, which in the age of `*ad_n` is tacitly assumed).

A transport may or may not provide any of these, and I don't think that EDHOC can be expected to cater for all these. It may be helpful to point out that for particular cases the CoAP transport provides blueprints (which would be straightforward to extend to both directions if a transport needs it), but IMO a generic wrapper would either be inflated out of proportion or killed by a thousand parameters.

**Comment by StefanHri** [comment]
What exactly is meant by correlation in section 3.4? Does this mean that the transport has some correlator field, e.g., CoAP tocken or something more?

I think the easiest way for building a generic library will be to pass not only the payload of the messages to the EDHOC engine but also the correlator from the transport if there is some. If there is no correlator field on the transport layer we can use the currently proposed method. In that case, the session mapping can be done completely within the edhoc library.

**Comment by chrysn** [comment]
As I understand message correlation, it just means that the transport needs to keep apart different exchanges if it supports them at all.

API-wise, I don't think that using C_x as correlators at the API surface is a good abstraction. Instead, the correlators at the EDHOC library should be dealt out by the EDHOC library uniformly. (In high-level languages that might be an object, in low level language it might be a numeric handle or a pointer, or even just the EDHOC state struct if the library user manage memory -- called `exchange` below). Code on the client side can then look like this:

```
exchange := new EDHOC-exchange(as initiator, forbidden_c_i=my_oscore_recipient_ids)
msg12 := new CoAP-exchange
msg12.send_request(True | exchange.build_message_1())
exchange.process_message_2(msg12.receive response())
msg34 := new CoAP-exchange
msg34.send_request(exchange.c_r | exchange.build_message_3())
```

which does just keeps the exchange handle on the stack, whereas on a server side there are actual lookups:

```
running_exchanges := new Dict mapping IntOrBstr to EDHOC-exchange
on request(req):
    c_r = first item of req
    if c_r is True:
        exchange := new EDHOC-exchange(
            as responder,
            forbidden_c_r=keys of running_exchanges ∪ my_oscore_recipient_ids,
            )
        exchange.process_message_1(later items of req)
        running_exchanges.store(exchange.c_r, exchange)
        return exchange.build_message_2()
    else:
        exchange := running_exchanges.load(c_r)
        exchange.process_message_3(later items of req)
        return exchange.build_message_4()
```

This keeps all the nitty-gritty transport details out of the EDHOC engine, which can now focus on running EDHOC only. (What this code is not showing is that ).

This is not ruling out EDHOC engines that manage all their contexts on their own, but for such engines to be truly generic as you envision them, they can't just index their exchanges by C_peer, but can also receive information from the transport indicating to which type of message is coming in, or through which transport, so that would need to be a more complex interface.

**Comment by StefanHri** [comment]
Actually this brings another aspect. We need to map messages to sessions only on the server side. Correct? On the client side this is not necessary since after a request we can receive only a response to the that request. Therefore we don't need to support different EDHOC sessions on the client side. 

Of course if an implementation supports several parallel running clients this can be the case. However, I don't know if there is a use case for that.

**Comment by StefanHri** [comment]
Do I understand correctly Christian that you will prefer a single concept to be used for correlation (an EDHOC session / handle ..) since this is what I am proposing too. However this will excised the 45 byte boundary that we have for message 2, see above.

**Comment by chrysn** [comment]
I prefer the transports to map to a single concept semantically (that is, the concept of individual EDHOC exchanges that might in APIs be expressed as objects or handles), but that mapping should still be done in implementations (guided by the transport specification).

On the wire, for the CoAP transport, I am very happy with the current construction -- it allows that mapping to happen easily inside the transport. Even when implementing both initiator and responder, both the client behavior (receive EDHOC messages, send peer C_x followed by EDHOC messages) and the server behavior (receive peer C_x, process rest of payload as EDHOC message; send plain EDHOC messages) are simple.

**Comment by StefanHri** [comment]
Does this mean that your implementation is tightly coupled to a CoAP transport and if you wish to add an additional transport you need to extend it?

I will say let's keep it like it is. I have now a sufficient understanding of how to implement it.

@stoprocent, @gselander should the pretending of C_x also be addressed in the test vectors?

**Comment by chrysn** [comment]
The implementation I work on is in two pieces: py-edhoc (which I'm extending based on Timothy's work) is unaware of any transports. aiocoap handles CoAP messages, prepends and reads C_x where necessary, selects keys and loads them depending on the CoAP library's configuration and properties of the incoming message (e.g. the Uri-Host). A new transport could use py-edhoc as is. (It could pick some parts about key selection from aiocoap --  I made no attempt to provide a generic key store so far).

**Comment by gselander** [comment]
@StefanHri About test vectors: Yes. Marek agreed that should be a feature of the new test vector code, and the prepended identifiers printed out in some vectors.

There is a meeting about test vectors on Monday April 4, 16:30-17:30 CEST, those interested to join can contact me.

**Comment by gselander** [comment]
Closing this as the final point about test vectors is included in #222.
